### ISSUE TITLE
Making requests with GPS points that are close together results in unexpected rounding of the matched points

### ISSUE THREAD

**Issue by Stefan781** [valhalla/valhalla#4190]
Sending a map matching request with GPS points that are near each other results in unexpected rounding in the response's matched points. An example would be:

{'shape': [{'lat': 12.345678, 'lon': 12.345678}, {'lat': 12.345678, 'lon': 12.345677}, {'lat': 12.345678, 'lon': 12.345678}...

The response then contains:

  "matched_points": [
    {
      "lon": 12.345678,
      "lat": 12.345678,
      "type": "matched",
      "edge_index": 0,
      "distance_along_edge": 0.419918,
      "distance_from_trace_point": 0.008194
    },
    {
      "lon": 12.345678, // ====>This one should end with a 7
      "lat": 12.345678, 
      "type": "interpolated",
      "edge_index": 0,
      "distance_along_edge": 0.419918,
      "distance_from_trace_point": 0.072941
    },
    {
      "lon": 12.345678,
      "lat": 12.345678,
      "type": "interpolated",
      "edge_index": 0,
      "distance_along_edge": 0.419918,
      "distance_from_trace_point": 0.008194
    }, ...

Is this the expected behavior? The distance_from_trace_point changes for the second point, which makes me think that internally the points are different, as it should be.

Is this, by any chance, related to the problem of matching overlapping points. If so, are there any plans to look into the issue?

**Comment by kevinkreiser** [comment]
im not so sure this has anything to do with the matching, its probably more to do with parsing/reserializing the request/response (since they start/end as strings and have to be coerced into/out of floating point types). we can write a simple unit test to see what happens after just parsing/serializing the request/response to proof/disprove this theory

**Comment by kevinkreiser** [comment]
sorry i mean inside of valhalla, we turn your input string into a floating point number, do stuff with it and then when done we turn it back into a string. it seems most likely to me that these conversions lose a 6th digit of precision but thats just a guess from years of dealing with this kind of annoyance :smile:

**Comment by Stefan781** [comment]
@kevinkreiser I tried sending a request with 5 digits after the decimal separator and got a response with 6 digits after the separator. Tried it with a request with 4 digits and the 4th digit of the response was changed + 2 more were added at the end to make it to 6.

**Comment by kevinkreiser** [comment]
yes the response is fixed to 6 digits regardless of what is input. can you maybe tell us what impact this has in your application?

**Comment by Stefan781** [comment]
I have managed to come up with a workaround for my use case for now, since the matched points correspond 1 to 1 with the input points. But perhaps there should be an entry in the documentation about the issue. A fix is always welcome, of course.

**Comment by kevinkreiser** [comment]
if you could say why you need an exact string match maybe we can help in a different way. we have an ID field which is a string and can be used to match a response with a request

**Comment by Stefan781** [comment]
Following the [issue with wrong edges](https://github.com/valhalla/valhalla/issues/4188), I removed duplicates as a workaround. However, I would still want to have all of the points matched, including the removed duplicates. The naive approach of using the lat and lon values resulted in the current issue. A workaround was to trace the duplicates internally and use the 1 to 1 mapping of input and output points by Valhalla instead of the lat and lon values. I do not see a use case for the ID. 

I could probably remove the duplicates altogether, which will fix the wrong edges in the first place. If you have other ideas - let me know. 

It would be useful to have the two issues mentioned somewhere so that it is easily searchable for future reference to other people who run into them, as it would have saved me some time with debugging my code. I apologize if this is already included somewhere.
